File: app/src/main/java/com/jajinba/pixabaydemo/presenter/ListPresenter.java
package com.jajinba.pixabaydemo.presenter;

import android.text.TextUtils;
import android.util.Log;

import com.jajinba.pixabaydemo.contract.ListContract;
import com.jajinba.pixabaydemo.model.ImageManager;
import com.jajinba.pixabaydemo.model.PixabayImageObject;
import com.jajinba.pixabaydemo.utils.ArrayUtils;

import java.util.ArrayList;
import java.util.List;

public class ListPresenter implements  ListContract.Presenter {
  private static final String TAG = ListPresenter.class.getSimpleName();

  private ListContract.View mView;

  public ListPresenter(ListContract.View view) {
    mView = view;

    ImageManager.getInstance().setCallback(new ImageManager.Callback() {
      @Override
      public void onSuccess(String keyword, List<PixabayImageObject> imageList) {
        Log.d(TAG, "onSuccess, image count:" + ArrayUtils.getLengthSafe(imageList));
        mView.searchFinished(keyword, imageList);
      }

      @Override
      public void onFailed(int errorMsg) {
        Log.d(TAG, "onFailed");
        mView.searchFinished(null, null);
      }
    });
  }

  @Override
  public List<PixabayImageObject> getImageList(String keyword) {
    if (TextUtils.isEmpty(keyword)) {
      return new ArrayList<>();
    } else {
      return ImageManager.getInstance().getImageListWithKeyword(keyword);
    }
  }

  @Override
  public void loadMore(final String keyword) {
    int searchPage = ImageManager.getInstance().previousLoadedPage(keyword) + 1;
    ImageManager.getInstance().searchImage(keyword, searchPage);
  }

  @Override
  public String getLastOperation() {
    return ImageManager.getInstance().getLastOperation();
  }

}

File: app/src/main/java/com/jajinba/pixabaydemo/presenter/ListPresenter.kt
package com.jajinba.pixabaydemo.presenter

import android.text.TextUtils
import android.util.Log
import com.jajinba.pixabaydemo.contract.ListContract
import com.jajinba.pixabaydemo.model.ImageManager
import com.jajinba.pixabaydemo.model.PixabayImageObject
import com.jajinba.pixabaydemo.utils.ArrayUtils
import java.util.*

class ListPresenter(private val mView: ListContract.View) : ListContract.Presenter {

  companion object {
    private val TAG = ListPresenter::class.java.simpleName
  }

  init {
    ImageManager.getInstance().setCallback(object : ImageManager.Callback {
      override fun onSuccess(keyword: String, imageList: MutableList<PixabayImageObject>) {
        Log.d(TAG, "onSuccess, image count:" + ArrayUtils.getLengthSafe(imageList))
//        mView.searchFinished(keyword, imageList)
      }

      override fun onFailed(errorMsg: Int) {
        Log.d(TAG, "onFailed")
//        mView.onFailed(errorMsg)
      }
    })
  }

  override fun getImageList(keyword: String): MutableList<PixabayImageObject> {
    Log.d(TAG, "mvp] getImageList")
    return if (TextUtils.isEmpty(keyword)) {
      ArrayList()
    } else {
      ImageManager.getInstance().getImageListWithKeyword(keyword)
    }
  }

  override fun loadMore(keyword: String) {
    Log.d(TAG, "mvp] loadMore")
    Log.d(TAG, "Load more images of $keyword")
    val searchPage: Int = ImageManager.getInstance().previousLoadedPage(keyword) + 1
//    ImageManager.getInstance().searchImage(keyword, searchPage)
  }

  override fun getLastOperation(): String {
    Log.d(TAG, "mvp] getLastOperation")
    return ImageManager.getInstance().getLastOperation() ?: ""
  }
}
